server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OffHeapStarTreeBuilder.java

```diff
@@ -0,0 +1,896 @@
+        int numBytes = 0;
+        // Initialize to set bits for dimensions which are null
+        byte dimensionNullBitSet = 0;
+        List<Byte> dimensionsByteBitSets = new ArrayList<>();
+        for (int i = 0; i < starTreeDocument.dimensions.length; i++) {
+            if (i % 8 == 0 && i > 0) {
+                dimensionsByteBitSets.add(dimensionNullBitSet);
+                dimensionNullBitSet = 0;
+            }
+            if (starTreeDocument.dimensions[i] == null) {
+                // Set the corresponding bit in dimensionNullBitSet to 1 (present)
+                dimensionNullBitSet |= (byte) (1 << (i % 8));
+                starTreeDocument.dimensions[i] = 0L;
+            }
+            output.writeLong(starTreeDocument.dimensions[i]);
+            numBytes += Long.BYTES;
+        }
+        dimensionsByteBitSets.add(dimensionNullBitSet);
+        for (Byte dimBitSet : dimensionsByteBitSets) {
+            output.writeByte(dimBitSet);
+            numBytes += Byte.BYTES;
+        }
+        return numBytes;
+    }
+
+    /**
+     * Write star tree document metrics to file
+     */
+    private int writeMetrics(StarTreeDocument starTreeDocument, IndexOutput output, boolean isAggregatedDoc) throws IOException {
+        int numBytes = 0;
+        List<Byte> byteBitSets = new ArrayList<>();
+        byte metricsNullBitSet = 0;
+        for (int i = 0; i < starTreeDocument.metrics.length; i++) {
+            if (i % 8 == 0 && i > 0) {
+                output.writeByte(metricsNullBitSet);
+                byteBitSets.add(metricsNullBitSet);
+                metricsNullBitSet = 0;
+            }
+            switch (metricAggregatorInfos.get(i).getValueAggregators().getAggregatedValueType()) {
+                case LONG:
+                    if (starTreeDocument.metrics[i] == null) {
+                        starTreeDocument.metrics[i] = 0L;
+                        metricsNullBitSet |= (byte) (1 << (i % 8));
+                    }
+                    output.writeLong((Long) starTreeDocument.metrics[i]);
+                    numBytes += Long.BYTES;
+                    break;
+                case DOUBLE:
+                    if (isAggregatedDoc) {
+                        if (starTreeDocument.metrics[i] == null) {
+                            starTreeDocument.metrics[i] = 0.0;
+                            metricsNullBitSet |= (byte) (1 << (i % 8));
+                        }
+                        long val = NumericUtils.doubleToSortableLong((Double) starTreeDocument.metrics[i]);
+                        output.writeLong(val);
+                        numBytes += Long.BYTES;
+                    } else {
+                        if (starTreeDocument.metrics[i] == null) {
+                            starTreeDocument.metrics[i] = 0L;
+                            metricsNullBitSet |= (byte) (1 << (i % 8));
+                        }
+                        output.writeLong((Long) starTreeDocument.metrics[i]);
```

Comment:
Why do we not need to use NumericUtils here? I may be missing something obvious here I feel.

Reply:
During flush , the values are already stored as long - ( during ingestion ) which is this case.

'isAggregatedDoc' case is common for both merge as well as when we read and write star tree aggregated documents during off heap build flow.

During that time metrics are in double format, so we had to convert it to long.